content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+## Background: Optimizing Code Layout
+
+Before we get too far in, let's consider how modern compilers approach code layout, specifically function placement. Since 1990, [Pettis and Hansen's code reordering method](https://dl.acm.org/citation.cfm?id=93550) has been the de facto approach, used by LLVM, HHVM, and other compiler infrastructures. Recent function layout techniques, such as call-chain clustering, attempt to improve upon the Pettis-Hansen method by addressing some of its limitations.
+
+### Pettis-Hansen Method
+Informally, given a program, the Pettis-Hansen method (hereafter referred to as PH for brevity) works as follows.
+
+1. Generate an undirected call graph, where nodes are functions and edges represent calls from functions to other functions. More formally, a function A and a function B would have an edge between each other if A calls B or B calls A.
+2. Run a benchmark workload on this program, keeping track of what calls are made.
+3. Generate a weighted call graph, where edge weights are proportional to the frequency of calls between the two corresponding functions.
+4. Go through the edges in decreasing weight, joining the nodes connected by the edge. When nodes join, their edges combine, summing weights when necessary. Node names are also concatenated, denoting the corresponding function order.
+    - The node names concatenate to maximize the sum of the weights between consecutive nodes. This can involve reversing node names.
+6. After this process is complete, no edges will remain in the graph, and the order in which nodes are joined denote the function order.
+
+Let's go through a brief example. Suppose the weighted call graph is as follows.
+
+<img src="https://i.imgur.com/tvee3J0.png" width="400px">
+
+
+From the above graph, we see that *A* calls *B* 100 times, *B* calls *C* 30 times, and so on. In the first iteration of the algorithm (step 4), we pick the edge with the largest weight. That's *AB*. We then combine nodes *A* and *B* to arrive at the following graph.
+
+<img src="https://i.imgur.com/cSvb1SL.png" width="400px">
+
+
+We then combine *C* and *D*.
+
+
+<img src="https://i.imgur.com/Ua3UW3I.png" width="400px">
+
+
+Here, we obviously combine *A-B* and *C-D*, but recall the rule about concatenating node names. There are four ways to coalesce the nodes:
+- *A-B-C-D*: Here, *BC* has a weight of 30.
+- *A-B-D-C*: *BD* has a weight of 0 (no edge), so this is worse than the above ordering.
+- *B-A-C-D*: *AC* has a weight of 90, which is now the best ordering.
+- *B-A-D-C*: *AD* has a weight of 0.
+
+Based on the above, *B-A-C-D* is the optimal ordering, since it maximizes the weight between the coalesced nodes.
+
+### Additional optimization by PH
+PH also conducts some basic block reordering to maximize spatial locality of the program. This is explored further when discusses Codestitcher's basic block chaining stage.
```

Comment:
In particular, it also does the thing where it splits a function _once_ into a "hot" part and a "cold" part, right? That seems like a big factor in PH that Codestitcher relaxes.

Specifically, this section seems to give the reader the impression that PH does not reorder individual basic blocks; it only works at the function granularity. That's not quite true.